docs/coalton-lisp-interop.md

```diff
@@ -0,0 +1,121 @@
+* Many optimizations are performed.
+
+### Pitfalls of Having Two Modes
+
+Unfortunately, having two modes may mean one inadvertently depend on the behavior in one mode that is not supported in the other. We advise testing your code against both modes.
+
+## Promises of Data and Basic Data Types
+
+**PROMISE**: Every Coalton value exists as a Lisp value. The structure of said value may not always be defined, however, and must be considered an element of type `T` in general circumstances.
+
+**PROMISE**: Coalton `Integer`, `Character`, `String`,  `Single-Float`, and `Double-Float` values correspond to their Lisp counterparts.
+
+**PROMISE**: `COALTON:Boolean` is a Lisp `BOOLEAN`; `COALTON:True` is Lisp `T`, and `COALTON:False` is Lisp `NIL`.
+
+**WARNING**: Lists and other data types may not correspond to Lisp's conception of them.
+
+## Promises of  `define-type`
+
+Consider a type:
+
+```
+(define-type (Foo ...)
+  (ctor1 ...)
+  ctor2
+  ...)
+```
+
+**PROMISE**: `FOO`, `FOO/CTOR1`, `FOO/CTOR2`, etc. will be valid symbols denoting types checkable by `typep`.
+
+**PROMISE**: `CTOR1`, `CTOR2`, etc. will all be bound lexical variables.
+
+**PROMISE**: Non-nullary constructors (e.g., `CTOR1`) will be Lisp functions.
+
+**PROMISE**: `(subtypep 'FOO/* 'FOO)` will be true for valid substitutions of `*`.
+
+**WARNING**: There is no promise whether these types are classes, or even if they are disjoint from other Lisp types. For example, in the current version of Coalton, `(define-type f a)` will emit a class `F/A` in development mode, and a type equivalent to `(member F/A)` in release mode. As such, one may **not** write `defmethod`s specializing on these types.
+
+**Todo?**: Promise when you have exactly one non-nullary constructor.
+
+### Achieving Guaranteed Lisp Compatibility with `(REPR :LISP)`
+
+The pragma `(repr :lisp)` helps achieve Lisp compatibility of structures regardless of interaction mode, at the expense of optimization opportunities. Consider a type:
+
+```
+(repr :lisp)
+(define-type (Foo ...)
+  (ctor1 ...)
+  ctor2
+  ...)
+```
+
+All of the promises of the previous section hold true, and additionally:
+
+**PROMISE**: `FOO` and each `FOO/*` each be classes and can be dispatched in a `defmethod`, disjoint from any other type in Lisp.
+
+**PROMISE**: Even if there is only a single constructor, a separate disjoint type will be created.
+
+**WARNING**: It is undefined whether these classes are `standard-class` or `structure-class` or something else.
+
+## Promises of `define`
+
+Consider the following definitions:
+
+```
+(define v       x1)
+(define (f ...) x2)
+```
+
+**PROMISE**: `v` and `f` will be lexical variables in Lisp bound to those symbols.
+
+**PROMISE**: `f` will be a function in Lisp of arity at the definition site.
+
+
+
+## Lisp-Calls-Coalton Bridge
+
+**Todo**: `COALTON` operator
+
+## Coalton-Calls-Lisp Bridge
+
+**Todo**: `LISP` operator
+
+## `LISP-VALUE` Type
+
+**Todo**
```

Comment:
Lisp-Object isn't a type at runtime. It get's compiled to `(declare (type t) ...)`. See https://github.com/coalton-lang/coalton/blob/913870e09e2776aeac4f55fd9696cd616f34faca/src/typechecker/environment.lisp#L150-L156.